File: ios/Plugin/CameraLayersPlugin.swift
import Foundation
import Capacitor
import AVFoundation
import Photos
import UIKit

@objc(CameraLayersPlugin)
public class CameraLayersPlugin: CAPPlugin {
    private var cameraView: LayeredCameraView?
    private var layers: [String: CameraLayerData] = [:]
    
    @objc func start(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            // Request camera permissions
            AVCaptureDevice.requestAccess(for: .video) { granted in
                if !granted {
                    call.reject("Camera permission denied")
                    return
                }
                
                DispatchQueue.main.async {
                    // Parse options
                    let position = call.getString("position") ?? "back"
                    let width = call.getDouble("width") ?? Double(UIScreen.main.bounds.width)
                    let height = call.getDouble("height") ?? 480.0
                    let x = call.getDouble("x") ?? 0.0
                    let y = call.getDouble("y") ?? 0.0
                    let toBack = call.getBool("toBack") ?? false
                    let enableZoom = call.getBool("enableZoom") ?? true
                    let enableFocus = call.getBool("enableFocus") ?? true
                    
                    // Create camera view
                    let frame = CGRect(x: x, y: y, width: width, height: height)
                    self.cameraView = LayeredCameraView(frame: frame)
                    
                    guard let cameraView = self.cameraView else {
                        call.reject("Failed to create camera view")
                        return
                    }
                    
                    // Configure camera
                    let cameraPosition: AVCaptureDevice.Position = position == "front" ? .front : .back
                    
                    // Add to webView
                    if let webView = self.bridge?.webView {
                        if toBack {
                            webView.superview?.insertSubview(cameraView, belowSubview: webView)
                        } else {
                            webView.superview?.addSubview(cameraView)
                        }
                    }
                    
                    // Start camera session
                    cameraView.startCamera(position: cameraPosition, enableZoom: enableZoom, enableFocus: enableFocus) { error in
                        if let error = error {
                            call.reject("Failed to start camera: \(error.localizedDescription)")
                        } else {
                            call.resolve()
                        }
                    }
                }
            }
        }
    }
    
    @objc func stop(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            self.cameraView?.stopCamera()
            self.cameraView?.removeFromSuperview()
            self.cameraView = nil
            call.resolve()
        }
    }
    
    @objc func capture(_ call: CAPPluginCall) {
        guard let cameraView = self.cameraView else {
            call.reject("Camera not started")
            return
        }
        
        let quality = call.getDouble("quality") ?? 85.0
        let outputType = call.getString("outputType") ?? "uri"
        let saveToGallery = call.getBool("saveToGallery") ?? false
        
        cameraView.capturePhoto(quality: quality / 100.0, layers: Array(self.layers.values)) { result, error in
            if let error = error {
                call.reject("Failed to capture photo: \(error.localizedDescription)")
                return
            }
            
            guard let result = result else {
                call.reject("Failed to capture photo")
                return
            }
            
            var response: [String: Any] = [
                "width": result.width,
                "height": result.height
            ]
            
            if outputType == "base64" || outputType == "both" {
                if let base64 = result.base64 {
                    response["base64"] = base64
                }
            }
            
            if outputType == "uri" || outputType == "both" {
                if let uri = result.uri {
                    response["uri"] = uri
                }
            }
            
            // Save to gallery if requested
            if saveToGallery, let image = result.image {
                PHPhotoLibrary.requestAuthorization { status in
                    if status == .authorized {
                        PHPhotoLibrary.shared().performChanges({
                            PHAssetCreationRequest.creationRequestForAsset(from: image)
                        }, completionHandler: nil)
                    }
                }
            }
            
            call.resolve(response)
        }
    }
    
    @objc func addLayer(_ call: CAPPluginCall) {
        guard let layerDict = call.getObject("layer") ?? call.options else {
            call.reject("Layer data required")
            return
        }
        
        let layerId = (layerDict["id"] as? String) ?? UUID().uuidString
        
        guard let layerData = self.parseLayer(layerId: layerId, dict: layerDict) else {
            call.reject("Invalid layer data")
            return
        }
        
        self.layers[layerId] = layerData
        
        DispatchQueue.main.async {
            self.cameraView?.updateLayers(Array(self.layers.values))
        }
        
        call.resolve(["layerId": layerId])
    }
    
    @objc func updateLayer(_ call: CAPPluginCall) {
        guard let layerId = call.getString("layerId") else {
            call.reject("Layer ID required")
            return
        }
        
        guard var existingLayer = self.layers[layerId] else {
            call.reject("Layer not found")
            return
        }
        
        if let layerDict = call.getObject("layer") {
            existingLayer = self.mergeLayer(existing: existingLayer, updates: layerDict)
            self.layers[layerId] = existingLayer
            
            DispatchQueue.main.async {
                self.cameraView?.updateLayers(Array(self.layers.values))
            }
        }
        
        call.resolve()
    }
    
    @objc func removeLayer(_ call: CAPPluginCall) {
        guard let layerId = call.getString("layerId") else {
            call.reject("Layer ID required")
            return
        }
        
        self.layers.removeValue(forKey: layerId)
        
        DispatchQueue.main.async {
            self.cameraView?.updateLayers(Array(self.layers.values))
        }
        
        call.resolve()
    }
    
    @objc func clearLayers(_ call: CAPPluginCall) {
        self.layers.removeAll()
        
        DispatchQueue.main.async {
            self.cameraView?.updateLayers([])
        }
        
        call.resolve()
    }
    
    @objc func switchCamera(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            self.cameraView?.switchCamera { error in
                if let error = error {
                    call.reject("Failed to switch camera: \(error.localizedDescription)")
                } else {
                    call.resolve()
                }
            }
        }
    }
    
    @objc func setFlashMode(_ call: CAPPluginCall) {
        guard let modeString = call.getString("mode") else {
            call.reject("Flash mode required")
            return
        }
        
        let flashMode: AVCaptureDevice.FlashMode
        switch modeString {
        case "on":
            flashMode = .on
        case "auto":
            flashMode = .auto
        case "off":
            flashMode = .off
        default:
            flashMode = .off
        }
        
        DispatchQueue.main.async {
            self.cameraView?.setFlashMode(flashMode)
            call.resolve()
        }
    }
    
    @objc func setZoom(_ call: CAPPluginCall) {
        guard let zoom = call.getDouble("zoom") else {
            call.reject("Zoom level required")
            return
        }
        
        DispatchQueue.main.async {
            self.cameraView?.setZoom(CGFloat(zoom))
            call.resolve()
        }
    }
    
    @objc func setFocus(_ call: CAPPluginCall) {
        guard let x = call.getDouble("x"), let y = call.getDouble("y") else {
            call.reject("Focus coordinates required")
            return
        }
        
        DispatchQueue.main.async {
            self.cameraView?.setFocus(point: CGPoint(x: x, y: y))
            call.resolve()
        }
    }
    
    // MARK: - Helper Methods
    
    private func parseLayer(layerId: String, dict: [String: Any]) -> CameraLayerData? {
        guard let type = dict["type"] as? String else { return nil }
        
        let x = (dict["x"] as? Double) ?? 0
        let y = (dict["y"] as? Double) ?? 0
        let width = dict["width"] as? Double
        let height = dict["height"] as? Double
        let zIndex = (dict["zIndex"] as? Int) ?? 0
        let opacity = (dict["opacity"] as? Double) ?? 1.0
        let rotation = (dict["rotation"] as? Double) ?? 0.0
        
        return CameraLayerData(
            id: layerId,
            type: type,
            x: CGFloat(x),
            y: CGFloat(y),
            width: width != nil ? CGFloat(width!) : nil,
            height: height != nil ? CGFloat(height!) : nil,
            zIndex: zIndex,
            opacity: CGFloat(opacity),
            rotation: CGFloat(rotation),
            imagePath: dict["imagePath"] as? String,
            imageUrl: dict["imageUrl"] as? String,
            text: dict["text"] as? String,
            fontSize: (dict["fontSize"] as? Double).map { CGFloat($0) },
            fontColor: dict["fontColor"] as? String,
            fontFamily: dict["fontFamily"] as? String,
            textAlign: dict["textAlign"] as? String,
            backgroundColor: dict["backgroundColor"] as? String,
            padding: (dict["padding"] as? Double).map { CGFloat($0) },
            shapeType: dict["shapeType"] as? String,
            strokeColor: dict["strokeColor"] as? String,
            strokeWidth: (dict["strokeWidth"] as? Double).map { CGFloat($0) },
            fillColor: dict["fillColor"] as? String
        )
    }
    
    private func mergeLayer(existing: CameraLayerData, updates: [String: Any]) -> CameraLayerData {
        return CameraLayerData(
            id: existing.id,
            type: (updates["type"] as? String) ?? existing.type,
            x: (updates["x"] as? Double).map { CGFloat($0) } ?? existing.x,
            y: (updates["y"] as? Double).map { CGFloat($0) } ?? existing.y,
            width: (updates["width"] as? Double).map { CGFloat($0) } ?? existing.width,
            height: (updates["height"] as? Double).map { CGFloat($0) } ?? existing.height,
            zIndex: (updates["zIndex"] as? Int) ?? existing.zIndex,
            opacity: (updates["opacity"] as? Double).map { CGFloat($0) } ?? existing.opacity,
            rotation: (updates["rotation"] as? Double).map { CGFloat($0) } ?? existing.rotation,
            imagePath: (updates["imagePath"] as? String) ?? existing.imagePath,
            imageUrl: (updates["imageUrl"] as? String) ?? existing.imageUrl,
            text: (updates["text"] as? String) ?? existing.text,
            fontSize: (updates["fontSize"] as? Double).map { CGFloat($0) } ?? existing.fontSize,
            fontColor: (updates["fontColor"] as? String) ?? existing.fontColor,
            fontFamily: (updates["fontFamily"] as? String) ?? existing.fontFamily,
            textAlign: (updates["textAlign"] as? String) ?? existing.textAlign,
            backgroundColor: (updates["backgroundColor"] as? String) ?? existing.backgroundColor,
            padding: (updates["padding"] as? Double).map { CGFloat($0) } ?? existing.padding,
            shapeType: (updates["shapeType"] as? String) ?? existing.shapeType,
            strokeColor: (updates["strokeColor"] as? String) ?? existing.strokeColor,
            strokeWidth: (updates["strokeWidth"] as? Double).map { CGFloat($0) } ?? existing.strokeWidth,
            fillColor: (updates["fillColor"] as? String) ?? existing.fillColor
        )
    }
}

// MARK: - Data Structures

struct CameraLayerData {
    let id: String
    let type: String
    let x: CGFloat
    let y: CGFloat
    let width: CGFloat?
    let height: CGFloat?
    let zIndex: Int
    let opacity: CGFloat
    let rotation: CGFloat
    
    // Image layer
    let imagePath: String?
    let imageUrl: String?
    
    // Text layer
    let text: String?
    let fontSize: CGFloat?
    let fontColor: String?
    let fontFamily: String?
    let textAlign: String?
    let backgroundColor: String?
    let padding: CGFloat?
    
    // Shape layer
    let shapeType: String?
    let strokeColor: String?
    let strokeWidth: CGFloat?
    let fillColor: String?
}

struct CaptureResultData {
    let image: UIImage?
    let base64: String?
    let uri: String?
    let width: Int
    let height: Int
}

File: android/src/main/java/com/kaelmo/cameralayers/CameraLayersPlugin.kt
package com.kaelmo.cameralayers

import android.Manifest
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.RectF
import android.util.Base64
import android.util.Log
import android.view.ViewGroup
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import com.getcapacitor.annotation.Permission
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.FileOutputStream
import java.util.*

@CapacitorPlugin(
    name = "CameraLayers",
    permissions = [
        Permission(strings = [Manifest.permission.CAMERA], alias = "camera"),
        Permission(strings = [Manifest.permission.WRITE_EXTERNAL_STORAGE], alias = "storage")
    ]
)
class CameraLayersPlugin : Plugin() {
    private var cameraView: LayeredCameraView? = null
    private val layers = mutableMapOf<String, CameraLayer>()

    @PluginMethod
    fun start(call: PluginCall) {
        if (!hasRequiredPermissions()) {
            requestAllPermissions(call, "cameraPermissionsCallback")
            return
        }

        startCameraInternal(call)
    }

    @PluginMethod
    fun stop(call: PluginCall) {
        bridge.executeOnMainThread {
            cameraView?.stopCamera()
            
            val parent = cameraView?.parent as? ViewGroup
            parent?.removeView(cameraView)
            
            cameraView = null
            call.resolve()
        }
    }

    @PluginMethod
    fun capture(call: PluginCall) {
        val cameraView = this.cameraView
        if (cameraView == null) {
            call.reject("Camera not started")
            return
        }

        val quality = call.getInt("quality", 85)
        val outputType = call.getString("outputType", "uri")
        val saveToGallery = call.getBoolean("saveToGallery", false)!!

        cameraView.capturePhoto(quality, layers.values.toList()) { result, error ->
            if (error != null) {
                call.reject("Failed to capture photo: ${error.message}")
                return@capturePhoto
            }

            if (result == null) {
                call.reject("Failed to capture photo")
                return@capturePhoto
            }

            val response = JSObject()
            response.put("width", result.width)
            response.put("height", result.height)

            when (outputType) {
                "base64" -> response.put("base64", result.base64)
                "uri" -> response.put("uri", result.uri)
                "both" -> {
                    response.put("base64", result.base64)
                    response.put("uri", result.uri)
                }
            }

            if (saveToGallery && result.bitmap != null) {
                // Save to gallery implementation
                saveToGallery(result.bitmap)
            }

            call.resolve(response)
        }
    }

    @PluginMethod
    fun addLayer(call: PluginCall) {
        val layerData = call.data
        val layerId = layerData.optString("id", UUID.randomUUID().toString())

        val layer = parseLayer(layerId, layerData) ?: run {
            call.reject("Invalid layer data")
            return
        }

        layers[layerId] = layer

        bridge.executeOnMainThread {
            cameraView?.updateLayers(layers.values.toList())
        }

        val result = JSObject()
        result.put("layerId", layerId)
        call.resolve(result)
    }

    @PluginMethod
    fun updateLayer(call: PluginCall) {
        val layerId = call.getString("layerId")
        if (layerId == null) {
            call.reject("Layer ID required")
            return
        }

        val existingLayer = layers[layerId]
        if (existingLayer == null) {
            call.reject("Layer not found")
            return
        }

        val updates = call.getObject("layer")
        val updatedLayer = mergeLayer(existingLayer, updates)
        layers[layerId] = updatedLayer

        bridge.executeOnMainThread {
            cameraView?.updateLayers(layers.values.toList())
        }

        call.resolve()
    }

    @PluginMethod
    fun removeLayer(call: PluginCall) {
        val layerId = call.getString("layerId")
        if (layerId == null) {
            call.reject("Layer ID required")
            return
        }

        layers.remove(layerId)

        bridge.executeOnMainThread {
            cameraView?.updateLayers(layers.values.toList())
        }

        call.resolve()
    }

    @PluginMethod
    fun clearLayers(call: PluginCall) {
        layers.clear()

        bridge.executeOnMainThread {
            cameraView?.updateLayers(emptyList())
        }

        call.resolve()
    }

    @PluginMethod
    fun switchCamera(call: PluginCall) {
        bridge.executeOnMainThread {
            cameraView?.switchCamera { error ->
                if (error != null) {
                    call.reject("Failed to switch camera: ${error.message}")
                } else {
                    call.resolve()
                }
            }
        }
    }

    @PluginMethod
    fun setFlashMode(call: PluginCall) {
        val mode = call.getString("mode", "off")

        bridge.executeOnMainThread {
            cameraView?.setFlashMode(mode)
            call.resolve()
        }
    }

    @PluginMethod
    fun setZoom(call: PluginCall) {
        val zoom = call.getFloat("zoom")
        if (zoom == null) {
            call.reject("Zoom level required")
            return
        }

        bridge.executeOnMainThread {
            cameraView?.setZoom(zoom)
            call.resolve()
        }
    }

    @PluginMethod
    fun setFocus(call: PluginCall) {
        val x = call.getFloat("x")
        val y = call.getFloat("y")

        if (x == null || y == null) {
            call.reject("Focus coordinates required")
            return
        }

        bridge.executeOnMainThread {
            cameraView?.setFocus(x, y)
            call.resolve()
        }
    }

    // Helper methods

    private fun startCameraInternal(call: PluginCall) {
        bridge.executeOnMainThread {
            val position = call.getString("position", "back")
            val width = call.getInt("width", bridge.webView.width)
            val height = call.getInt("height", 480)
            val x = call.getInt("x", 0)
            val y = call.getInt("y", 0)
            val toBack = call.getBoolean("toBack", false)!!
            val enableZoom = call.getBoolean("enableZoom", true)!!
            val enableFocus = call.getBoolean("enableFocus", true)!!

            try {
                val view = LayeredCameraView(context)
                cameraView = view

                val layoutParams = ViewGroup.LayoutParams(width, height)
                view.layoutParams = layoutParams
                view.x = x.toFloat()
                view.y = y.toFloat()

                val parent = bridge.webView.parent as? ViewGroup
                if (parent != null) {
                    if (toBack) {
                        parent.addView(view, 0)
                    } else {
                        parent.addView(view)
                    }
                }

                val cameraFacing = if (position == "front") 
                    LayeredCameraView.CAMERA_FACING_FRONT 
                else 
                    LayeredCameraView.CAMERA_FACING_BACK

                view.startCamera(cameraFacing, enableZoom, enableFocus) { error ->
                    if (error != null) {
                        call.reject("Failed to start camera: ${error.message}")
                    } else {
                        call.resolve()
                    }
                }
            } catch (e: Exception) {
                call.reject("Failed to create camera view: ${e.message}")
            }
        }
    }

    private fun parseLayer(layerId: String, data: JSObject): CameraLayer? {
        val type = data.optString("type") ?: return null

        return CameraLayer(
            id = layerId,
            type = type,
            x = data.optDouble("x", 0.0).toFloat(),
            y = data.optDouble("y", 0.0).toFloat(),
            width = if (data.has("width")) data.optDouble("width").toFloat() else null,
            height = if (data.has("height")) data.optDouble("height").toFloat() else null,
            zIndex = data.optInt("zIndex", 0),
            opacity = data.optDouble("opacity", 1.0).toFloat(),
            rotation = data.optDouble("rotation", 0.0).toFloat(),
            imagePath = data.optString("imagePath"),
            imageUrl = data.optString("imageUrl"),
            text = data.optString("text"),
            fontSize = if (data.has("fontSize")) data.optDouble("fontSize").toFloat() else null,
            fontColor = data.optString("fontColor"),
            fontFamily = data.optString("fontFamily"),
            textAlign = data.optString("textAlign"),
            backgroundColor = data.optString("backgroundColor"),
            padding = if (data.has("padding")) data.optDouble("padding").toFloat() else null,
            shapeType = data.optString("shapeType"),
            strokeColor = data.optString("strokeColor"),
            strokeWidth = if (data.has("strokeWidth")) data.optDouble("strokeWidth").toFloat() else null,
            fillColor = data.optString("fillColor")
        )
    }

    private fun mergeLayer(existing: CameraLayer, updates: JSObject): CameraLayer {
        return CameraLayer(
            id = existing.id,
            type = updates.optString("type", existing.type),
            x = updates.optDouble("x", existing.x.toDouble()).toFloat(),
            y = updates.optDouble("y", existing.y.toDouble()).toFloat(),
            width = if (updates.has("width")) updates.optDouble("width").toFloat() else existing.width,
            height = if (updates.has("height")) updates.optDouble("height").toFloat() else existing.height,
            zIndex = updates.optInt("zIndex", existing.zIndex),
            opacity = updates.optDouble("opacity", existing.opacity.toDouble()).toFloat(),
            rotation = updates.optDouble("rotation", existing.rotation.toDouble()).toFloat(),
            imagePath = updates.optString("imagePath", existing.imagePath),
            imageUrl = updates.optString("imageUrl", existing.imageUrl),
            text = updates.optString("text", existing.text),
            fontSize = if (updates.has("fontSize")) updates.optDouble("fontSize").toFloat() else existing.fontSize,
            fontColor = updates.optString("fontColor", existing.fontColor),
            fontFamily = updates.optString("fontFamily", existing.fontFamily),
            textAlign = updates.optString("textAlign", existing.textAlign),
            backgroundColor = updates.optString("backgroundColor", existing.backgroundColor),
            padding = if (updates.has("padding")) updates.optDouble("padding").toFloat() else existing.padding,
            shapeType = updates.optString("shapeType", existing.shapeType),
            strokeColor = updates.optString("strokeColor", existing.strokeColor),
            strokeWidth = if (updates.has("strokeWidth")) updates.optDouble("strokeWidth").toFloat() else existing.strokeWidth,
            fillColor = updates.optString("fillColor", existing.fillColor)
        )
    }

    private fun saveToGallery(bitmap: Bitmap) {
        // Implementation to save bitmap to gallery
        // This would use MediaStore on Android
    }

    private fun hasRequiredPermissions(): Boolean {
        return hasPermission(Manifest.permission.CAMERA)
    }
}

// Data classes

data class CameraLayer(
    val id: String,
    val type: String,
    val x: Float,
    val y: Float,
    val width: Float?,
    val height: Float?,
    val zIndex: Int,
    val opacity: Float,
    val rotation: Float,
    val imagePath: String?,
    val imageUrl: String?,
    val text: String?,
    val fontSize: Float?,
    val fontColor: String?,
    val fontFamily: String?,
    val textAlign: String?,
    val backgroundColor: String?,
    val padding: Float?,
    val shapeType: String?,
    val strokeColor: String?,
    val strokeWidth: Float?,
    val fillColor: String?
)

data class CaptureResult(
    val bitmap: Bitmap?,
    val base64: String?,
    val uri: String?,
    val width: Int,
    val height: Int
)
